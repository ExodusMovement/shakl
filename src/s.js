import React from 'react';

import flatten from './flatten';

const s = (Comp, { name, multi, ...opts } = {}) => style => {
  const S = React.forwardRef(({ children, ...props }, ref) => {
    const { attrs = {}, comp, child } = opts;

    const styles = multi ? style : { style };

    return React.createElement(
      comp || Comp,
      {
        ref,
        ...attrs,
        ...props,
        ...Object.keys(styles).reduce(
          (obj, prop) => ({
            ...obj,
            [prop]: flatten([
              attrs[prop],
              typeof styles[prop] === 'function'
                ? styles[prop](props)
                : styles[prop],
              props[prop]
            ])
          }),
          {}
        )
      },
      child ? React.createElement(child, {}, children) : children
    );
  });

  S.displayName = name || `styled(${Comp.displayName || Comp.name})`;

  S.extend = more => s(S, { name })(more);
  S.withProps = attrs => s(S, { name, attrs })();
  S.withComponent = comp => s(S, { name, comp })(style);
  S.withChild = child => s(S, { name, child })();

  return S;
};

export default s;
