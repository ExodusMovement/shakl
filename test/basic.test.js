import React from 'react'

import { create as r } from 'react-test-renderer'

import { View } from 'react-native'

import s from '../src'

it('creates a styled component', () => {
  const Foo = s.View({ flex: 1 })
  const foo = r(<Foo />).toJSON()

  expect(foo.props.style).toEqual({ flex: 1 })

  expect(foo).toMatchSnapshot()
})

it('creates a styled component with null static styles', () => {
  const Foo = s.View()
  const foo = r(<Foo />).toJSON()

  expect(foo.props.style).toEqual({})

  expect(foo).toMatchSnapshot()
})

it('creates a styled component with null dynamic styles', () => {
  const Foo = s.View(() => null)
  const foo = r(<Foo />).toJSON()

  expect(foo.props.style).toEqual({})

  expect(foo).toMatchSnapshot()
})

it('creates a styled component with empty static styles', () => {
  const Foo = s.View({})
  const foo = r(<Foo />).toJSON()

  expect(foo.props.style).toEqual({})

  expect(foo).toMatchSnapshot()
})

it('creates a styled component with empty dynamic styles', () => {
  const Foo = s.View(() => ({}))
  const foo = r(<Foo />).toJSON()

  expect(foo.props.style).toEqual({})

  expect(foo).toMatchSnapshot()
})

it('creates a styled component with dynamic styles based on props', () => {
  const Foo = s.View((p) => ({ padding: p.padded ? 10 : 0 }))

  const foo = r(<Foo />).toJSON()
  const fooPadded = r(<Foo padded />).toJSON()

  expect(foo.props.style).toEqual({ padding: 0 })
  expect(fooPadded.props.style).toEqual({ padding: 10 })

  expect(foo).toMatchSnapshot()
  expect(fooPadded).toMatchSnapshot()
})

it('creates a styled component of a functional component', () => {
  const Foo = (props) => <View {...props} />
  const Bar = s(Foo)({ flex: 1 })

  const foo = r(<Foo />).toJSON()
  const bar = r(<Bar />).toJSON()

  expect(bar.props.style).toEqual({ flex: 1 })

  expect(foo).toMatchSnapshot()
  expect(bar).toMatchSnapshot()
})

it('creates a styled component of a class component', () => {
  class Foo extends React.Component {
    render() {
      return <View {...this.props} />
    }
  }

  const Bar = s(Foo)({ flex: 1 })

  const foo = r(<Foo />).toJSON()
  const bar = r(<Bar />).toJSON()

  expect(bar.props.style).toEqual({ flex: 1 })

  expect(foo).toMatchSnapshot()
  expect(bar).toMatchSnapshot()
})
