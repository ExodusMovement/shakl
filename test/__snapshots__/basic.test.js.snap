// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a styled component 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`creates a styled component of a class component 1`] = `<View />`;

exports[`creates a styled component of a class component 2`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`creates a styled component of a functional component 1`] = `<View />`;

exports[`creates a styled component of a functional component 2`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`creates a styled component with combined static and dynamic styles 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "padding": 0,
    }
  }
/>
`;

exports[`creates a styled component with combined static and dynamic styles 2`] = `
<View
  padded={true}
  style={
    Object {
      "flex": 1,
      "padding": 10,
    }
  }
/>
`;

exports[`creates a styled component with dynamic styles based on props 1`] = `
<View
  style={
    Object {
      "padding": 0,
    }
  }
/>
`;

exports[`creates a styled component with dynamic styles based on props 2`] = `
<View
  padded={true}
  style={
    Object {
      "padding": 10,
    }
  }
/>
`;

exports[`creates a styled component with empty dynamic styles 1`] = `
<View
  style={Object {}}
/>
`;

exports[`creates a styled component with empty static styles 1`] = `
<View
  style={Object {}}
/>
`;

exports[`creates a styled component with multiple static style objects 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
/>
`;

exports[`creates a styled component with null dynamic styles 1`] = `
<View
  style={Object {}}
/>
`;

exports[`creates a styled component with null static styles 1`] = `
<View
  style={Object {}}
/>
`;
